### PR TITLE
cli: additional enhancements after switching to jackc/pgx 

### LINKED ISSUE #72065: cli: `cockroach sql` may override `DateStyle` when set with `-c`
When running this test:

```
spawn $argv sql --url "postgresql://test@localhost:26257?options=-cdatestyle%3Dymd"
eexpect root@
send "SHOW datestyle;\r"
eexpect "ISO, YMD"
eexpect root@
interrupt
eexpect eof
```

with `cockroach sql`, sometimes the `DateStyle` is `YMD`, sometimes the `DateStyle` is ISO,MDY.

This is because lib/pq sends ISO,MDY along as well:

https://github.com/lib/pq/blob/26399a76874b6cf27f551d02b9576e77b211ead1/connector.go#L89-L96

And as it is assigned to a map in the snippet below, the last one wins.

https://github.com/cockroachdb/cockroach/blob/12afa456ab273fdb640d35ea3b5c676a4474e914/pkg/sql/pgwire/server.go#L850-L868

i imagine this to be a similar problem with other fields

Jira issue: CRDB-10931

### PR DESCRIPTION
**Pull request by rafiss**
Fixes https://github.com/cockroachdb/cockroach/issues/72065

Release note (cli change): Using COPY in the CLI is now
supported while inside an explicit transaction.

Release note (bug fix): The DateStyle session setting is no longer
ignored when using the CLI when set in the `options` URL parameter.

Release note (cli change): Ctrl+C (the interrupt signal) can now be used
in the CLI to attempt to cancel the currently executing SQL query.

### CODE REVIEW

**Comment by knz** [comment]
i'd like to delay the review of this until the other PR has merged, because if there's a chance for a rebase that will make reviewable unhappy.

or alternatively, you can create a PR using the other PR as base branch, to facilitate the review. (The way I did for e.g. #82069.)

**Comment by rafiss** [comment]
> or alternatively, you can create a PR using the other PR as base branch, to facilitate the review. (The way I did for e.g. https://github.com/cockroachdb/cockroach/pull/82069.)

That's what I intended to do here. Anyway, the other PR should get merged soon, so this will look better soon.

File: pkg/cli/clisqlshell/sql.go
Comment (otan) [inline pkg/cli/clisqlshell/sql.go:2207]: was there a test for this functionality somewhere?

File: pkg/cli/clisqlclient/rows.go
Comment (otan) [inline pkg/cli/clisqlclient/rows.go]: you mean the function?

**Review by otan — APPROVED** [review]
what a nice library pgx is :')

File: pkg/cli/clisqlclient/rows.go
Comment (otan) [reply pkg/cli/clisqlclient/rows.go]: (gets deleted later)

File: pkg/cli/clisqlshell/sql.go
Comment (rafiss) [reply pkg/cli/clisqlshell/sql.go:2207]: done

File: pkg/cli/clisqlshell/sql.go
Comment (rafiss) [reply pkg/cli/clisqlshell/sql.go:1860]: done

**Review by otan — APPROVED** [review]

**Review by knz — APPROVED** [review]
:lgtm: modulo missing release note.

also the interrupt test has an issue in CI? need to investigate further maybe.

> Reviewed 1 of 3 files at r2, 1 of 59 files at r9, 19 of 19 files at r16, 2 of 2 files at r17, 1 of 1 files at r18, 6 of 6 files at r19, 13 of 13 files at r20, 3 of 3 files at r21, 7 of 7 files at r22, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82101#-:-N3Z3WxAD1ukkSLghiRI:blzbftv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @knz, @otan, and @rafiss)

___
*[`-- commits` line 12 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/82101#-N3Z2SZt97GPvF5UJ0oa):*
Maybe add a release note to explain that query cancellation is enabled in the CLI.

**Comment by rafiss** [comment]
> also the interrupt test has an issue in CI? need to investigate further maybe.

🤦  looks like pgx has the same default behavior where cancelling the context actually closes the connection. it's defined by the context watcher here: https://github.com/jackc/pgconn/blob/7ddbd74d5e5a52cd24f750cacfc9be91d4f49f76/pgconn.go#L360

**Comment by rafiss** [comment]
I've updated things so it explicitly sends a cancel request instead of relying on context cancellation. i know this wasn't your first preference, so let me know what you think

**Review by rafiss — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82101)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @knz, @otan, and @rafiss)

___
*[`pkg/cli/clisqlshell/sql.go` line 2238 at r28](https://reviewable.io/reviews/cockroachdb/cockroach/82101#-N3aM4ab5UJHssjewIGU:-N3aM4ab5UJHssjewIGV:b8gts88) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab7c3118c54a5041bb6ad55af800664852f1e25d/pkg/cli/clisqlshell/sql.go#L2238)):*
> ```Go
> 						break wait
> 					case <-tooLongTimer:
> 						fmt.Fprintln(c.iCtx.stderr, "server does not respond to query cancellation; a second interrupt will stop the shell.")
> ```

an alternative approach is to call the `cancel()` function (the one from line 2204 above) here. this would cause the connection to be closed, rather than exiting the shell.

**Review by rafiss — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82101)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @knz, @otan, and @rafiss)

___
*[`pkg/cli/clisqlshell/sql.go` line 2238 at r28](https://reviewable.io/reviews/cockroachdb/cockroach/82101#-N3aM4ab5UJHssjewIGU:-N3aMQYvCoL8m1cvMVR4:b-xnb59c) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab7c3118c54a5041bb6ad55af800664852f1e25d/pkg/cli/clisqlshell/sql.go#L2238)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

an alternative approach is to call the `cancel()` function (the one from line 2204 above) here. this would cause the connection to be closed, rather than exiting the shell.
</blockquote></details>

oh nvm, that is not correct. we would have to call the `cancel()` function that's created in `runWithInterruptableCtx`

**Comment by rafiss** [comment]
@knz ping in case you had any thoughts on this:
> I've updated things so it explicitly sends a cancel request instead of relying on context cancellation. i know this wasn't your first preference, so let me know what you think

**Review by knz — APPROVED** [review]
This looks good with a tweak.

> Reviewed 1 of 20 files at r23, 1 of 1 files at r24, 2 of 6 files at r25, 7 of 13 files at r26, 2 of 3 files at r27, 7 of 7 files at r28, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82101#-:-N3tMwDd1lSPJ47lLiLn:b-ymvvy7)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @knz, @otan, and @rafiss)

___
*[`pkg/cli/clisqlshell/context.go` line 79 at r28](https://reviewable.io/reviews/cockroachdb/cockroach/82101#-N3tMluG-9asi4-OF1s4:-N3tMluG-9asi4-OF1s5:buvsxwo) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab7c3118c54a5041bb6ad55af800664852f1e25d/pkg/cli/clisqlshell/context.go#L79)):*
> ```Go
> 	mu struct {
> 		syncutil.Mutex
> 		cancelFn func() error
> ```

can you make that take a context.Context argument
___
*[`pkg/cli/clisqlshell/sql.go` line 2221 at r28](https://reviewable.io/reviews/cockroachdb/cockroach/82101#-N3tMp9TDwx-QM1X3KGR:-N3tMp9TDwx-QM1X3KGS:baxwtqn) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab7c3118c54a5041bb6ad55af800664852f1e25d/pkg/cli/clisqlshell/sql.go#L2221)):*
> ```Go
> 				// Cancel the query's context, which should make the driver
> 				// send a cancellation message.
> 				if err := cancelFn(); err != nil {
> ```

then pass ctx as argument here
___
*[`pkg/cli/clisqlshell/sql.go` line 2268 at r28](https://reviewable.io/reviews/cockroachdb/cockroach/82101#-N3tMqxP7H8boOg2LNNx:-N3tMqxP7H8boOg2LNNy:b-5qynz) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab7c3118c54a5041bb6ad55af800664852f1e25d/pkg/cli/clisqlshell/sql.go#L2268)):*
> ```Go
> 	// Inform the Ctrl+C handler that this query is executing.
> 	c.iCtx.mu.Lock()
> 	c.iCtx.mu.cancelFn = func() error {
> ```

then simplify this `.cancel = c.conn.Cancel`

**Review by rafiss — COMMENTED** [review]
tftr! i've added your suggestion

bors r=knz,otan

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82101#-:-N3uqxDh9wWXqy3Z02to:bmw1kwi)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @knz, @otan, and @rafiss)

**Pull request merged**
